src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -0,0 +1,217 @@
+
+/**
+ * Class that monitors consumer groups consuming from topics; registers its metrics with a metric registry.
+ *
+ * Runs a thread that polls the Kafka broker; use {@link #start()} and {@link #stop()} to spin up and down.
+ *
+ * The metric namespace is as follows.
+ *
+ * <p>
+ * {@code <metric prefix>.<topic>.{partition.%d,total}.{size,offset,lag}}
+ *
+ * <p>
+ * Size is the maximum offset of the topic (this is independent of any consumer). Offset is the current offset of the
+ * consumer. Lag is the size minus the offset. Totals are summed across the partitions.
+ * NB: Since the queries for the topic sizes and consumer offsets are necessarily separate, they race; this means that
+ * lag can possibly be negative.
+ *
+ * <p>
+ * Future work: Have alternate constructor in which you don't specify any topics, and the class uses the
+ * {@link OffsetMonitor} to dynamically register metrics based on the topics that the consumer group is consuming.
+ *
+ * <p>
+ * Future work: Have alternate constructor in which you don't specify any groups either, and the class uses the
+ * {@link OffsetMonitor} to dynamically register metrics based on all consumer groups' topic consumption.
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public class OffsetMetrics implements Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(OffsetMetrics.class);
+
+    private final String metricPrefix;
+    private final MetricRegistry metricRegistry;
+    private final String groupId;
+    private final Collection<String> topics;
+    private final Duration pollPeriod;
+    private final OffsetMonitor monitor;
+    private final Map<String, Metric> metricMap;
+    private final ScheduledExecutorService exec;
+
+    public OffsetMetrics(
+            final String metricPrefix,
+            final MetricRegistry metricRegistry,
+            final String groupId,
+            final String brokerList,
+            final Collection<String> topics) {
+        this(metricPrefix, metricRegistry, groupId, brokerList, topics, Duration.ofSeconds(10));
+    }
+
+    @VisibleForTesting
+    OffsetMetrics(
+            final String metricPrefix,
+            final MetricRegistry metricRegistry,
+            final String groupId,
+            final String brokerList,
+            final Collection<String> topics,
+            final Duration pollPeriod) {
+        Preconditions.checkArgument(metricPrefix != null, "null metric prefix");
+        Preconditions.checkArgument(!topics.isEmpty(), "no topics");
+        this.metricPrefix = metricPrefix;
+        this.metricRegistry = metricRegistry;
+        this.groupId = groupId;
+        this.topics = topics;
+        this.pollPeriod = pollPeriod;
+        monitor = new OffsetMonitor(groupId, brokerList);
+
+        final Collection<String> badTopics = new ArrayList<>();
+        final ImmutableMap.Builder<String, Metric> builder = ImmutableMap.builder();
+        topics.forEach(topic -> {
```

Comment:
In the ad attribution service we used a Histogram metric to aggregate the lag for the different partitions. It gives you things like max/min/mean/stddev for free. The code is very simple
```
    private void calculateLag(ConsumerRecords<byte[], AttributionABEvent> records) {
        Set<TopicPartition> partitions = records.partitions();
        Map<TopicPartition, Long> offsetMap = this.consumer.endOffsets(partitions);
        for (TopicPartition tp : partitions) {
            long currentOffset = this.consumer.position(tp);
            long maxOffset = offsetMap.get(tp);
            long lag = maxOffset - currentOffset;
            histogram.update(lag);
        }
    }

```

The downside of this approach is that it is done synchronously inside the poll loop so having something that does this "on the side" would be ideal. Would you consider adding a histogram metric as well? We used both a 5 minute SlidingTimeWindowReservoir and an ExponentiallyDecayingReservoir but didn't find much of a difference in practice.

Reply:
The other downside of your approach is that you can't track per-partition lag exactly.  However, I do really like the idea of using a histogram for easier-to-consume summary statistics.

> Would you consider adding a histogram metric as well?

Absolutely!  Let me give it a shot; I'll keep in mind your comment about your experience with the different backing reservoirs.  Based on my experience, the non-exponentially decaying reservoir would be most useful for very low-rate metrics.

Reply:
Check out 662ed908550d750d8d725b2c1a2d26361459ae3b...7f9782049ba5d063dce6b055c7764942d9594c5e @enorrbinot.

Some uses here:
- https://github.com/opentable/service-agassoc/commit/5dcfac6b7713dd86c969bcb4aac7a5eb92f60ed4
- https://github.com/opentable/service-chat/commit/1a000f7a8faf756e86d3c84d343f58ba3252410c#diff-f142017d86b14a9c22244aaa2e93ecd1

Reply:
Use PRs:

- https://github.com/opentable/service-chat/pull/95
- https://github.com/opentable/service-agassoc/pull/22

Reply:
LGTM, we should try this out in ad attribution.

Reply:
Baller.  Just waiting on final review from @stevenschlansker; then I'll release everything and update the parent.  I'll let you know.